make SDK.jl async

dehann opened the pull request:
close
- #85 

This PR makes async the calls for all function similar to SDK.py, but the Diana.jl query and mutate internal calls are still blocking -- we can easily add tnose two asyncs here in SDK if Diana.jl is not doing them internally already.  Since query and mutate as lower level calls, they are usually called by user API functions that already include the async layer, so okay for time being until we know more about Diana.jl

~~@jim-hill-r , do you know of any other functions that should be made async?  If not then SDK.jl should be ready for v0.4.0.  Ill do the release and tagging with confirmation on any missing async functions.~~

dehann commented:
alright, I have a better idea of where to add the asyncs here.  I'm going to expand this PR to look a bit more like SDK.py before merging and going into release reviews.

dehann commented:
Going to try finish in the afternoon, but may drag into the later evening -- rather fix it properly and close out SDK.jl v0.4.0 more inline with the other SDKs

dehann commented:
Hi Sam, this should be a near complete async equivalent to SDK.py.  Feel free to merge, or let me know,  I'll merge as soon I hear back from you our Jim.  Will test / update Tut1 and if all good, with request final review on SDK.jl main before we register v0.4.0

jim-hill-r reviewed (approved):
Looks good. Only note would be style related. The functions preceded by underscore is bad style, IMHO, as it doesn't convey intent. I would have just named those functions 'solveSessionSync' and not exported them. Just a style thing for water-cooler debate.

GearsAD commented on src/navability/entities/NavAbilityClient.jl:
```diff
@@ -24,9 +24,11 @@ end
 function NavAbilityHttpsClient(apiUrl::String="https://api.d1.navability.io")::NavAbilityClient
     dianaClient = GraphQLClient(apiUrl)
     function query(options::QueryOptions)
+        # TODO does Diana or Deloittes return async?
```
Looks like the Deloitte one is synchronous as well - https://github.com/DeloitteDigitalAPAC/GraphQLClient.jl, so we'd need to do the async wrapping here

dehann replied:
Okay, I'll add that in a next PR.

GearsAD commented on src/navability/services/Factor.jl:
```diff
@@ -55,7 +55,9 @@ function getFactor(navAbilityClient::NavAbilityClient, client::Client, label::St
-function getFactors(navAbilityClient::NavAbilityClient, client::Client; detail::QueryDetail = SKELETON)::Vector{Dict{String,Any}}
+getFactor(navAbilityClient::NavAbilityClient, client::Client, label::String) = @async _getFactor(navAbilityClient, client, label)
+
+function _getFactors(navAbilityClient::NavAbilityClient, client::Client; detail::QueryDetail = SKELETON)::Vector{Dict{String,Any}}
```
I quite like the wrapping way of doing it

GearsAD commented on src/navability/services/Utils.jl:
```diff
@@ -10,18 +10,18 @@ Args:
-    maxSeconds::Int = 60,
-    expectedStatuses::Vector{String} = Nothing,
-    exceptionMessage::String = "Requests did not complete in time")
+    requestIds::AbstractVector{<:AbstractString};
```
Thanks

GearsAD commented on src/navability/services/Utils.jl:
```diff
@@ -31,4 +31,14 @@ function waitForCompletion(
+# Dispatch for vector of Tasks
+function waitForCompletion(
+    navAbilityClient::NavAbilityClient,
+    requestIds::AbstractVector{<:Task};
```
I need to try this, curious how the task thing works

dehann replied:
In Python when you do `async def` you get a `<coroutine object...>` back.  In Julia `@async` gives a `Task` object back.  Its the same thing.  The Julia one is also fully consolidated with mutlithreaded tasks via dispatch.

dehann commented:
I also updated the signatures away from `_` and added adjective `Event` instead.  These are all private functions, so should not make too big a difference and we can fix in patch release as necessary, no deprecation cycle required for private functions.
